src/ops/indexing.rs

```diff
@@ -0,0 +1,1791 @@
+        axes.push(i as i32);
+        operation_len += 1;
+        match op {
+            TakeIndex { index } => {
+                let item = Array::from_int(resolve_index_unchecked(
+                    *index,
+                    src.dim(i as i32) as usize,
+                ) as i32);
+                gather_indices.push(Rc::new(item));
+                is_slice.push(false);
+            }
+            Slice(range) => {
+                slice_count += 1;
+                is_slice.push(true);
+
+                let size = shape[i];
+                let absolute_start = range.absolute_start(size);
+                let absolute_end = range.absolute_end(size);
+                let indices = absolute_indices(absolute_start, absolute_end, range.stride());
+
+                let item = Array::from_slice(&indices, &[indices.len() as i32]);
+
+                gather_indices.push(Rc::new(item));
+            }
+            TakeArray { indices } => {
+                is_slice.push(false);
+                max_dims = max_dims.max(indices.ndim());
+                gather_indices.push(Rc::clone(indices));
+            }
+            Ellipsis | ExpandDims => {
+                unreachable!("Unexpected operation in gather_nd")
+            }
+        }
+    }
+
+    // reshape them so that the int/array indices are first
+    if gather_first {
+        if slice_count > 0 {
+            let mut slice_index = 0;
+            for (i, item) in gather_indices.iter_mut().enumerate() {
+                if is_slice[i] {
+                    let mut new_shape = vec![1; max_dims + slice_count];
+                    new_shape[max_dims + slice_index] = item.dim(0);
+                    *item = Rc::new(item.reshape(&new_shape));
+                    slice_index += 1;
+                } else {
+                    let mut new_shape = item.shape().to_vec();
+                    new_shape.extend((0..slice_count).map(|_| 1));
+                    *item = Rc::new(item.reshape(&new_shape));
+                }
+            }
+        }
+    } else {
+        // reshape them so that the int/array indices are last
+        for (i, item) in gather_indices[..slice_count].iter_mut().enumerate() {
+            let mut new_shape = vec![1; max_dims + slice_count];
+            new_shape[i] = item.dim(0);
+            *item = Rc::new(item.reshape(&new_shape));
+        }
+    }
+
+    // Do the gather
+    // let indices = new_mlx_vector_array(gather_indices);
+    // SAFETY: indices will be freed at the end of this function. The lifetime of the items in
+    // `gather_indices` is managed by the `gather_indices` vector.
+    let indices = unsafe {
+        let c_vec_array = mlx_sys::mlx_vector_array_new();
+        for item in gather_indices.iter() {
+            mlx_sys::mlx_vector_array_add(c_vec_array, item.c_array);
+        }
+        c_vec_array
+    };
+    let mut slice_sizes = shape.to_vec();
+    (0..operation_len).for_each(|i| slice_sizes[i] = 1);
```

Comment:
I've noticed you've been trying to optimize by reusing the for in operations loop - you can also leverage that for setting these initial values.

Reply:
Addressed in b7a0ff51933e54218df2811f3cdc1b916cadae5c